### DISCUSSION TITLE
`cirrus run` pulls macOS image every run for some reason

### DISCUSSION THREAD

**Discussion by mhanberg** [cirruslabs/cirrus-cli#645] [Q&A] [answered]
on my M2 mac mini that I use as a CI box, this seems to work as I would expect. One costly run to pull the image, then it's a noop after that.

Running it on my local m1 macbook air to test some tasks, it seems to pull it every time, which makes it takes 10 minutes.

Is there anything that I'm doing wrong?

Screenshot shows a successful run, then restarting it and see it pulling again.

<img width="1015" alt="CleanShot 2023-09-08 at 10 04 49@2x" src="https://github.com/cirruslabs/cirrus-cli/assets/5523984/c86a1f4c-ef70-46f2-99d4-61cfd82ce531">

**Comment by edigaryev** [comment]
Hello Mitchell 👋 

>Is there anything that I'm doing wrong?

1. How much free disk space do you have on average before/after pulling the image?

2. How many different remote Tart VM images do you use? It's possible that Tart's garbage collection mechanism kicks in because you don't have enough disk space to store more than one remote OCI image in Tart's cache.

3. Can you reproduce this with only using `tart`? What would be the steps to reproduce your issue?

**Comment by mhanberg** [comment]
> How much free disk space do you have on average before/after pulling the image?

before: about 65gigs
during: like 20ish i think (will double check later)
after: about 65 gigs

> How many different remote Tart VM images do you use? It's possible that Tart's garbage collection mechanism kicks in because you don't have enough disk space to store more than one remote OCI image in Tart's cache.

I only have ever tried with one

```yaml
macos_instance:
    image: ghcr.io/cirruslabs/macos-ventura-base:latest
```

> Can you reproduce this with only using tart? What would be the steps to reproduce your issue?

one time i did `tart pull ghcr.io/cirruslabs/macos-ventura-base:latest` and then did `cirrus run`, and it skipped the pull, but then after it was done, the disk space had returned to ~65gb and the image was gone it seems (i'm not sure where it stores them locally)

**Reply by edigaryev** [reply]
> one time i did `tart pull ghcr.io/cirruslabs/macos-ventura-base:latest` and then did `cirrus run`, and it skipped the pull, but then after it was done, the disk space had returned to ~65gb and the image was gone it seems (i'm not sure where it stores them locally)

Which Tart version are you using?

**Reply by mhanberg** [reply]
tart: 1.12.0
cirrus: cirrus version 0.102.0-945fd44

i installed them both via homebrew last weekend

**Reply by edigaryev** [reply, marked as answer]
Could you please run `export TART_NO_AUTO_PRUNE=true` and try again in the same terminal session?

It should disable the automatic cache pruning functionality [we have in `tart clone`](https://github.com/cirruslabs/tart/blob/4bb248e7b4395e1c1d4683053b710ff6b87f92d8/Sources/tart/Commands/Clone.swift#L66-L69), but it might result in errors when your VM allocates more real space on the disk because you don't seem to have enough space for two allocated VMs of that type.

**Reply by mhanberg** [reply]
I'll try that this weekend, just closed up shop for the day. Thanks!

**Reply by mhanberg** [reply]
this is the output

```
└─  TART_NO_AUTO_PRUNE=true cirrus run --tart-lazy-pull  --output simple -e PACKAGES="fd elixir" -e URL="https://github.com/neovim/neovim/releases/download/v0.8.3/nvim-macos.tar.gz"
Started 'Test' Task
Started 'pull virtual machine'
'pull virtual machine' skipped in 0.0s!
Started 'clone virtual machine'
Cloning virtual machine ghcr.io/cirruslabs/macos-ventura-base:latest...
pulling manifest...
pulling disk (16.7 GB compressed)...
0%
```

**Reply by mhanberg** [reply]
running a second time, it seemed to do what i expected

```
│elixir-tools.nvim on  main +29 -2 [$!?⇣] via 🌙 v5.4.6 took 9m37s
└─  TART_NO_AUTO_PRUNE=true cirrus run --tart-lazy-pull  --output simple -e PACKAGES="fd elixir" -e URL="https://github.com/neovim/neovim/releases/download/v0.8.3/nvim-macos.tar.gz"
Started 'Test' Task
Started 'pull virtual machine'
'pull virtual machine' skipped in 0.0s!
Started 'clone virtual machine'
Cloning virtual machine ghcr.io/cirruslabs/macos-ventura-base:latest...
'clone virtual machine' succeeded in 0.1s!
Started 'boot virtual machine'
VM was assigned with 192.168.65.4 IP
'boot virtual machine' succeeded in 1.0s!
Started 'syncing working directory'
'syncing working directory' succeeded in 15s!
```

now my laptop has 12G remaining storage

**Reply by mhanberg** [reply]
now it has 6G

**Reply by mhanberg** [reply]
once it finished (in about 1 min), storage went back to 12G
